pkg/bdb/bdb.go

```diff
@@ -0,0 +1,150 @@
+package bdb
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var validPageSizes = map[uint32]struct{}{
+	512:   {},
+	1024:  {},
+	2048:  {},
+	4096:  {},
+	8192:  {},
+	16384: {},
+	32768: {},
+	65536: {},
+}
+
+type BerkeleyDB struct {
+	file         *os.File
+	HashMetadata *HashMetadataPage
+}
+
+type Entry struct {
+	Value []byte
+	Err   error
+}
+
+func Open(path string) (*BerkeleyDB, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+
+	// read just a bit in to parse at least the metadata...
+	metadataBuff := make([]byte, 512)
+	_, err = file.Read(metadataBuff)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read metadata: %w", err)
+	}
+
+	_, err = file.Seek(0, io.SeekStart)
+	if err != nil {
+		return nil, fmt.Errorf("failed to seek db file: %w", err)
+	}
+
+	hashMetadata, err := ParseHashMetadataPage(metadataBuff)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, ok := validPageSizes[hashMetadata.PageSize]; !ok {
+		return nil, fmt.Errorf("unexpected page size: %+v", hashMetadata.PageSize)
+	}
+
+	return &BerkeleyDB{
+		file:         file,
+		HashMetadata: hashMetadata,
+	}, nil
+
+}
+
+func (db *BerkeleyDB) Read() <-chan Entry {
+	entries := make(chan Entry)
+
+	go func() {
+		defer close(entries)
+
+		// the first content entry (idx=0) is the db metadata, skip to the first real entry and keep reading content values
```

Comment:
nit: It looks like `HashPageType` is accepted and others are skipped. It's ok since the first entry is also not `HashPageType`. But this comment seems to skip the first entry with specific logic. I didn't find it.

Reply:
If I remember right, the first entry will always be a generic metadata page and never a hash metadata page, but I also can't seem to locate where in the code this is indicated directly. It's safe enough to process the first page anyway since we explicitly check the pageType and skip if it is not of interest. 

I'll change this to start processing at index 0 instead.

Reply:
Yeah, I also think it's safe enough even if it starts at index 0. Thanks.